### PR TITLE
Display indices of Polyhedron_item primitives in Polyhedron_demo

### PR DESCRIPTION
**Pull request by maxGimeno**
This PR introduces a new feature in the Polyhedron_demo.
- I+left click toggles the Id of the pointed primitive
- Ctrl+I toggles the Ids of all the primitives.

### CODE REVIEW

File: Polyhedron/demo/Polyhedron/Scene.cpp [hunk 1]
```diff
@@ -1002,7 +1002,23 @@ float Scene::get_bbox_length() const
     return bbox().height();
 }
 
+void Scene::printPrimitiveId(QPoint point, CGAL::Three::Viewer_interface* viewer)
+{
+    //TO DO :: Add a if condition to disable this when the printId mode is off
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Scene.cpp]: Is this todo still valid?

Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/Scene.cpp]: No, it has been done, bu the branch does not work under Windows, for some
reason, so don't merge it yet.

2015-12-17 18:01 GMT+01:00 Sebastien Loriot <notifications@github.com>:

**Comment by sloriot** [comment]
There are some issues on windows that need to be fixed.

**Comment by maxGimeno** [comment]
I tested it on my windows 7, it works, so it seems to be ready to be merged.

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-85

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-86

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 2]
```diff
@@ -410,8 +420,14 @@ Scene_polyhedron_item::initialize_buffers(CGAL::Three::Viewer_interface* viewer)
     normals_gouraud.resize(0);
     std::vector<float>(normals_gouraud).swap(normals_gouraud);
     are_buffers_filled = true;
+
+    CGAL::set_halfedgeds_items_id(*poly);
+
+	if (viewer->hasText())
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: Indentation issue

File: Polyhedron/demo/Polyhedron/Viewer.cpp [hunk 3]
```diff
@@ -1057,3 +1099,100 @@ void Viewer::wheelEvent(QWheelEvent* e)
...
+bool Viewer::textDisplayed() const
+{
+    return has_text;
+}
+void TextRenderer::draw(CGAL::Three::Viewer_interface *viewer)
+{
+    QPainter *painter = viewer->painter;
+	if (!painter->isActive())
+		painter->begin(viewer);
+    QRect rect;
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Viewer.cpp]: Indentation issue

File: Polyhedron/demo/Polyhedron/Viewer.cpp [hunk 4]
```diff
@@ -1057,3 +1099,100 @@ void Viewer::wheelEvent(QWheelEvent* e)
...
+
+ void TextRenderer::addText(float p_x, float p_y, float p_z, QString p_text, QFont p_font , QColor p_color )
+ {
+     local_textItems.append(new TextItem(p_x, p_y, p_z, p_text, p_font, p_color));
+ }
+
+ void TextRenderer::removeText(TextItem *item)
+ {
+             local_textItems.removeAll(item);
+ }
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Viewer.cpp]: Indentation issue

File: Polyhedron/demo/Polyhedron/Viewer.h [hunk 5]
```diff
@@ -79,6 +84,8 @@ public Q_SLOTS:
                                float animation_duration = 0.5f);
 
 protected:
+	void paintEvent(QPaintEvent *);
+    void paintGL();
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Viewer.h]: Indentation

File: Polyhedron/demo/Polyhedron/Viewer.h [hunk 6]
```diff
@@ -132,4 +145,74 @@ public Q_SLOTS:
...
+    void addText(float p_x, float p_y, float p_z, QString p_text, QFont font = QFont(), QColor p_color = Qt::black);
+    void removeText(TextItem*);
+    void removeTextList(TextListItem*);
+    QList<TextItem*> local_textItems;
+    QList<TextListItem*> items() const{return textItems;}
+    void setScene(CGAL::Three::Scene_interface* p_scene){scene = p_scene;}
+private:
+    QList<TextListItem*> textItems;
+    CGAL::Three::Scene_interface *scene;
+
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Viewer.h]: The member `scene` does not seem used.

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 7]
```diff
@@ -595,12 +604,13 @@ Scene_polyhedron_item::Scene_polyhedron_item()
...
       plugin_has_set_color_vector_m(false)
 {
-   // setItemIsMulticolor(true);
     cur_shading=FlatPlusEdges;
     is_selected = true;
     nb_facets = 0;
     nb_lines = 0;
+    textItems = new TextListItem(this);
     init();
+	targeted_id = NULL;
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: indentation

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 8]
```diff
@@ -613,13 +623,14 @@ Scene_polyhedron_item::Scene_polyhedron_item(Polyhedron* const p)
...
-   // setItemIsMulticolor(true);
     cur_shading=FlatPlusEdges;
     is_selected = true;
     nb_facets = 0;
     nb_lines = 0;
+    textItems = new TextListItem(this);
     init();
     invalidate_buffers();
+	targeted_id = NULL;
 }
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: indentation

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 9]
```diff
@@ -631,19 +642,34 @@ Scene_polyhedron_item::Scene_polyhedron_item(const Polyhedron& p)
...
-    //setItemIsMulticolor(true);
     cur_shading=FlatPlusEdges;
     is_selected=true;
+    textItems = new TextListItem(this);
     init();
     nb_facets = 0;
     nb_lines = 0;
     invalidate_buffers();
+	targeted_id = NULL;
 }
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: indentation

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 10]
```diff
@@ -631,19 +642,34 @@ Scene_polyhedron_item::Scene_polyhedron_item(const Polyhedron& p)
...
+      CGAL::Three::Viewer_interface* v = qobject_cast<CGAL::Three::Viewer_interface*>(viewer);
+
+      //Clears the targeted Id
+      targeted_id = NULL;
+      delete targeted_id;
+      //Remove textitems
+      v->textRenderer->removeTextList(textItems);
+      textItems =0;
+      delete textItems;
+    }
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: There is a memory leak, here: line 670 calls `delete 0`, which does nothing. It should be:
```C++
delete textItems;
textItems = 0;
```
and `textItems = 0;` is useless in a destructor: as the object is being destroyed, the value of `textItems` does no long matter.

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 11]
```diff
@@ -631,19 +642,34 @@ Scene_polyhedron_item::Scene_polyhedron_item(const Polyhedron& p)
...
     delete poly;
+    QGLViewer* viewer = *QGLViewer::QGLViewerPool().begin();
+    if(viewer)
+    {
+      CGAL::Three::Viewer_interface* v = qobject_cast<CGAL::Three::Viewer_interface*>(viewer);
+
+      //Clears the targeted Id
+      targeted_id = NULL;
+      delete targeted_id;
+      //Remove textitems
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: Memory leak! (`delete NULL`)

File: Three/include/CGAL/Three/Viewer_interface.h [hunk 12]
```diff
@@ -43,7 +44,12 @@ class VIEWER_EXPORT Viewer_interface : public QGLViewer, public QOpenGLFunctions
   Q_OBJECT
 
 public:
-  enum OpenGL_program_IDs { PROGRAM_WITH_LIGHT,
+ QPainter *painter;
+ TextRenderer *textRenderer;
+ //!Returns true if the primitive ids are displayed
```
Comment (lrineau) [inline Three/include/CGAL/Three/Viewer_interface.h]: `Viewer_interface` is an *interface*. It should be an empty class, with pure virtual methods, and nothing else!

File: Polyhedron/demo/Polyhedron/Viewer.cpp [hunk 13]
```diff
@@ -65,11 +71,14 @@ Viewer::Viewer(QWidget* parent, bool antialiasing)
...
   axis_are_displayed = true;
+  has_text = false;
+  i_is_pressed = false;
 }
 
 Viewer::~Viewer()
 {
   delete d;
+  delete painter;
 }
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Viewer.cpp]: All data members of `Viewer` should be hidden in the `d` pointer. That avoids to expose implementation details in `Viewer.h`.

For explanations of d-pointers, see:
- https://wiki.qt.io/D-Pointer
- https://en.wikipedia.org/wiki/Opaque_pointer

That is a common implementation trick, that I discovered first in the source code of Qt4.

File: Polyhedron/demo/Polyhedron/Viewer.h [hunk 14]
```diff
@@ -100,11 +109,15 @@ public Q_SLOTS:
   std::vector<float> c_Axis;
   //! Decides if the axis system must be drawn or not
   bool axis_are_displayed;
+  //! Decides if the text is displayed in the drawVidulaHints function.
+  bool has_text;
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Viewer.h]: is `drawVidulaHints` what you wanted to write or a typo?

**Comment by sloriot** [comment]
Comments from @lrineau seems to have been addressed

File: Polyhedron/demo/Polyhedron/Viewer.h (hunk 14 above)
Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/Viewer.h]: Definitely a typo !

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-100

**Comment by maxGimeno** [comment]
The conflicts have been eradicated. This PR is ready to be merged.

File: Polyhedron_IO/include/CGAL/IO/Polyhedron_scan_OFF.h [hunk 15]
```diff
@@ -114,6 +114,7 @@ Polyhedron_scan_OFF<HDS>:: operator()( HDS& target) {
             scanner.scan_facet_vertex_index( index, i);
             B.add_vertex_to_facet( index);
         }
+        //TO DO : Insert read color
```
Comment (sgiraudot) [inline Polyhedron_IO/include/CGAL/IO/Polyhedron_scan_OFF.h]: Was this TODO supposed to be done before this pull request or not?

**Comment by sgiraudot** [comment]
I tested it. The display looks good on screen, however:

* When the polyhedron is big, displaying all the indices with _Ctrl+i_ takes a lot of time. With a 40000 facets Polyhedron, the demo freezes for about 4 seconds on my computer. I initially tried with a 900000 facets Polyhedron and had to kill the demo as it was obviously too long. Wouldn't it be a good idea to add an upper-bound on the size of the polyhedron that prevents indices from being displayed if the user 
* _i + left click_ generates a segfault if no item is selected.
* More generally, the demo segfaults in many occasions because of the index display. For example, if I do _i + left click_ on a polyhedron, delete the item and load another polyhedron, the demo segfaults. Same if I display indices on several polyhedra and then delete one of them. My guess is that the deletion of text items is not handled correctly.
* Still with _i + left click_, very weird behavior if I display indices on several items: it would make sense that the indices are displayed _if and only if_ their related polyhedron is displayed, but this is not the case. It seems that all indices (from all polyhedra) are displayed if the current item is a _displayed_ polyhedron.
* The behavior of _Ctrl + i_ with several items is: the indices are only displayed for the selected polyhedron. I don't really mind this one, this is a perfectly acceptable behavior, but it would be nice to have a way to know if the _Ctrl + i_ option is ON or OFF (why not add a line to the _View_ menu like _Draw two sides_) as it is a global option of the viewer and that it doesn't apply to non-selected vertices. Also, when displaying the indices of a polyhedron that overlaps another one, the indices "hidden" by the facets of the other polyhedron are not hidden (it is not a big problem but it gives surprising displays).

**Comment by afabri** [comment]
I agree that an upper bound makes sense. 
It would also be good if when  I want the index of a vertex, this displays the indices in the area, say the 1-ring.   And  for an edge at the same time the indices of two vertices and the two faces.

**Comment by maxGimeno** [comment]
Ready for merging

**Comment by bo0ts** [comment]
@sloriot asked me to do a quick review. I think having the function `printPrimitiveId` in `Scene_item` is OK for now. Unfortunately this has several draw backs:

- there is no pretty way to check whether a derived class implements this functionality
- the complexity of the interface increases

My main concern is that this has to be done for almost every feature one would want to add to a specific derived class of `Scene_item`. What is needed is some kind of capabilities system where a derived class announces the kind of features it supports.

Another option would be to move this operation to a plug-in and let the plug-in check that it only works on a `Scene_polyhedron_item` via `applicable`.

**Comment by lrineau** [comment]
@sloriot @bo0ts @maxGimeno 

We could use an interface (a pure virtual base class): an item type with the feature "print something" would derive from it, and others would not. The viewer could then try a `dynamic_cast` to check which items have the feature.

That way, the interface `Scene_item` would stay as-is.

**Comment by maxGimeno** [comment]
@lrineau 
I like this idea.

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-121

**Comment by lrineau** [comment]
That is true that we need a collaboration of the `Scene` class. Let me change my proposal:

> We could use an interface (a pure virtual base class): an item type with the feature "print something" would derive from it, and others would not. The **scene** could then try a `dynamic_cast` to check which items have the feature.
> That way, the interface `Scene_item` would stay as-is.

**Comment by maxGimeno** [comment]
I rebased the branch. I tested it locally but it would be better to have some other developers to test it too.

**Comment by sloriot** [comment]
* [ ] `Max text items displayed %1` -> Set Max Numbers of Id Displayed" ?
* [ ] The label of the pop-up window to set the max is `QInputDialog::getText()`
* [ ] Take two models, select ids on one of them, then hide the other one. The id disappears.

**Comment by maxGimeno** [comment]
@sloriot 
- The thing is a text item can be used for other things that displaying Ids, for example I will use it to display a distance between to selected points, so it doesn't really makes sense to call the option "Set max Number  of Id Displayed". It could be "Set Maximum Number of Text Items Displayed" though.
- The ids do not disappear, it is simply that only the ids of the selected item are displayed (to avoid confusion), and when you hide the other one, you select it. If you want your id back, just select the first item again. It does not work correctly though, I have to fix something.

**Comment by maxGimeno** [comment]
Fixed

**Pull request merged**